core/unit_test/UnitTest_ScopeGuard.cpp

```diff
@@ -0,0 +1,150 @@
+ * Test to create a scope guard normally.
+ */
+TEST_F(scope_guard, create) {
+  // run it in a different process so side effects are not kept
+  EXPECT_EXIT(
+      {
+        {
+          Kokkos::ScopeGuard guard{};
+
+          if (!Kokkos::is_initialized()) std::exit(EXIT_FAILURE);
+          if (Kokkos::is_finalized()) std::exit(EXIT_FAILURE);
+        }
+
+        if (Kokkos::is_initialized()) std::exit(EXIT_FAILURE);
+        if (!Kokkos::is_finalized()) std::exit(EXIT_FAILURE);
+
+        std::exit(EXIT_SUCCESS);
+      },
+      testing::ExitedWithCode(EXIT_SUCCESS), "");
+}
+
+/**
+ * Test to create a scope guard with an argument.
+ */
+TEST_F(scope_guard, create_argument) {
+  // run it in a different process so side effects are not kept
+  EXPECT_EXIT(
+      {
+        {
+          Kokkos::InitializationSettings settings{};
+          Kokkos::ScopeGuard guard{settings};
+        }
+
+        std::exit(EXIT_SUCCESS);
+      },
+      testing::ExitedWithCode(EXIT_SUCCESS), "");
+}
+
+/**
+ * Test to create another scope guard when one has been created.
+ */
+TEST_F(scope_guard_DeathTest, create_while_initialize) {
+  ::testing::FLAGS_gtest_death_test_style = "threadsafe";
+  EXPECT_DEATH(
+      {
+        Kokkos::ScopeGuard guard1{};
+
+        // create a second scope guard while there is one already existing
+        Kokkos::ScopeGuard guard2{};
+      },
+      "Creating a ScopeGuard while Kokkos is initialized");
+}
+
+/**
+ * Test to create a scope guard when initialization has been done manually.
+ */
+TEST_F(scope_guard_DeathTest, create_after_initialize) {
+  ::testing::FLAGS_gtest_death_test_style = "threadsafe";
+  EXPECT_DEATH(
+      {
+        Kokkos::initialize();
+
+        // create a scope guard after manual initialization
+        Kokkos::ScopeGuard guard{};
+      },
+      "Creating a ScopeGuard while Kokkos is initialized");
+}
+
+/**
+ * Test to create another scope guard when one has been destroyed.
+ */
+TEST_F(scope_guard_DeathTest, create_after_finalize) {
+  ::testing::FLAGS_gtest_death_test_style = "threadsafe";
+  EXPECT_DEATH(
+      {
+        { Kokkos::ScopeGuard guard1{}; }
+
+        // create a second scope guard while the first one has been destroyed
+        // already
+        Kokkos::ScopeGuard guard2{};
+      },
+      "Creating a ScopeGuard after Kokkos was finalized");
+}
+
+/**
+ * Test to destroy a scope guard when finalization has been done manually.
+ */
+TEST_F(scope_guard_DeathTest, destroy_after_finalize) {
+  ::testing::FLAGS_gtest_death_test_style = "threadsafe";
+  EXPECT_DEATH(
+      {
+        // create a scope guard and finalize it manually
+        Kokkos::ScopeGuard guard{};
+        Kokkos::finalize();
+      },
+      "Destroying a ScopeGuard after Kokkos was finalized");
+}
+
+/**
+ * Static tests
+ */
+
+// Test scope guard is not copyable.
+static_assert(!std::is_copy_assignable<Kokkos::ScopeGuard>());
+static_assert(!std::is_copy_constructible<Kokkos::ScopeGuard>());
```

Comment:
Please also test that it is not moveable.

Reply:
I'm wondering if those `static_asserts` should be in the code, not the test (theoretically, one could have a reference counted `ScopeGuard` with reasonable copy semantics).

Reply:
I'll add the assert to this aim, but I'm not sure to be able to tell where it should be.

Reply:
Having them in the tests where they currently are is fine.

Reply:
Ok.